File: src/main/java/org/moinex/ui/dialog/creditcard/CreditCardInvoicePaymentController.java
/*
 * Filename: CreditCardInvoicePaymentController.java
 * Created on: October 30, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.dialog.creditcard;

import jakarta.persistence.EntityNotFoundException;
import java.math.BigDecimal;
import java.time.YearMonth;
import java.util.List;
import javafx.fxml.FXML;
import javafx.scene.control.ComboBox;
import javafx.scene.control.Label;
import javafx.scene.control.TextField;
import javafx.stage.Stage;
import lombok.NoArgsConstructor;
import org.moinex.error.MoinexException;
import org.moinex.model.creditcard.CreditCard;
import org.moinex.model.creditcard.CreditCardPayment;
import org.moinex.model.wallettransaction.Wallet;
import org.moinex.service.CalculatorService;
import org.moinex.service.CreditCardService;
import org.moinex.service.I18nService;
import org.moinex.service.WalletService;
import org.moinex.ui.common.CalculatorController;
import org.moinex.util.Constants;
import org.moinex.util.UIUtils;
import org.moinex.util.WindowUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.stereotype.Controller;

/**
 * Controller for the Credit card invoice payment dialog
 */
@Controller
@NoArgsConstructor
public class CreditCardInvoicePaymentController {
    @FXML private Label crcNameLabel;

    @FXML private Label crcInvoiceDueLabel;

    @FXML private Label crcInvoiceMonthLabel;

    @FXML private Label crcAvailableRebateLabel;

    @FXML private Label walletAfterBalanceLabel;

    @FXML private Label walletCurrentBalanceLabel;

    @FXML private Label totalToPayLabel;

    @FXML private ComboBox<Wallet> walletComboBox;

    @FXML private TextField useRebateValueField;

    private ConfigurableApplicationContext springContext;

    private WalletService walletService;

    private CreditCardService creditCardService;

    private CalculatorService calculatorService;

    private I18nService i18nService;

    private List<Wallet> wallets;

    private CreditCard creditCard = null;

    private YearMonth invoiceDate = null;

    /**
     * Constructor
     * @param walletService WalletService
     * @param creditCardService CreditCardService
     * @param calculatorService CalculatorService
     * @param i18nService I18nService
     * @note This constructor is used for dependency injection
     */
    @Autowired
    public CreditCardInvoicePaymentController(
            WalletService walletService,
            CreditCardService creditCardService,
            CalculatorService calculatorService,
            ConfigurableApplicationContext springContext,
            I18nService i18nService) {
        this.walletService = walletService;
        this.creditCardService = creditCardService;
        this.calculatorService = calculatorService;
        this.springContext = springContext;
        this.i18nService = i18nService;
    }

    public void setCreditCard(CreditCard crc, YearMonth invoiceDate) {
        this.creditCard = crc;
        this.invoiceDate = invoiceDate;
        configureComboBoxes();

        crcNameLabel.setText(creditCard.getName());

        BigDecimal invoiceAmount =
                creditCardService
                        .getPendingCreditCardPayments(
                                creditCard.getId(),
                                invoiceDate.getMonthValue(),
                                invoiceDate.getYear())
                        .stream()
                        .map(CreditCardPayment::getAmount)
                        .reduce(BigDecimal.ZERO, BigDecimal::add);

        crcInvoiceDueLabel.setText(UIUtils.formatCurrency(invoiceAmount));

        totalToPayLabel.setText(UIUtils.formatCurrency(invoiceAmount));

        crcInvoiceMonthLabel.setText(UIUtils.formatShortMonthYear(invoiceDate, i18nService));

        crcAvailableRebateLabel.setText(UIUtils.formatCurrency(creditCard.getAvailableRebate()));

        if (creditCard.getAvailableRebate().compareTo(BigDecimal.ZERO) > 0) {
            crcAvailableRebateLabel.setStyle("-fx-text-fill: green;");
        } else {
            crcAvailableRebateLabel.setStyle("-fx-text-fill: black;");
        }

        if (creditCard.getDefaultBillingWallet() != null) {
            walletComboBox.setValue(creditCard.getDefaultBillingWallet());
            UIUtils.updateWalletBalanceLabelStyle(
                    walletComboBox.getValue(), walletCurrentBalanceLabel);
            walletAfterBalance();
        }
    }

    @FXML
    private void initialize() {
        configureComboBoxes();

        loadWalletsFromDatabase();

        populateWalletComboBox();

        // Reset all labels
        UIUtils.resetLabel(walletAfterBalanceLabel);
        UIUtils.resetLabel(walletCurrentBalanceLabel);
        UIUtils.resetLabel(crcNameLabel);
        UIUtils.resetLabel(crcInvoiceDueLabel);
        UIUtils.resetLabel(crcInvoiceMonthLabel);

        walletComboBox.setOnAction(
                e -> {
                    UIUtils.updateWalletBalanceLabelStyle(
                            walletComboBox.getValue(), walletCurrentBalanceLabel);
                    walletAfterBalance();
                });

        // Update wallet after balance when the value field changes
        useRebateValueField
                .textProperty()
                .addListener(
                        (observable, oldValue, newValue) -> {
                            if (!newValue.matches(Constants.MONETARY_VALUE_REGEX)) {
                                useRebateValueField.setText(oldValue);
                            } else {
                                walletAfterBalance();
                            }
                        });
    }

    @FXML
    private void handleCancel() {
        Stage stage = (Stage) crcNameLabel.getScene().getWindow();
        stage.close();
    }

    @FXML
    private void handleSave() {
        Wallet wallet = walletComboBox.getValue();

        if (wallet == null) {
            WindowUtils.showInformationDialog(
                    i18nService.tr(
                            Constants.TranslationKeys.CREDITCARD_DIALOG_WALLET_NOT_SELECTED_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .CREDITCARD_DIALOG_WALLET_NOT_SELECTED_MESSAGE));
            return;
        }

        BigDecimal invoiceAmount =
                creditCardService
                        .getPendingCreditCardPayments(
                                creditCard.getId(),
                                invoiceDate.getMonthValue(),
                                invoiceDate.getYear())
                        .stream()
                        .map(CreditCardPayment::getAmount)
                        .reduce(BigDecimal.ZERO, BigDecimal::add);

        BigDecimal rebateValue =
                useRebateValueField.getText().isEmpty()
                        ? BigDecimal.ZERO
                        : new BigDecimal(useRebateValueField.getText());

        if (invoiceAmount.compareTo(BigDecimal.ZERO) == 0) {
            WindowUtils.showInformationDialog(
                    i18nService.tr(
                            Constants.TranslationKeys.CREDITCARD_DIALOG_INVOICE_ALREADY_PAID_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .CREDITCARD_DIALOG_INVOICE_ALREADY_PAID_MESSAGE));
        } else {
            try {
                creditCardService.payInvoice(
                        creditCard.getId(),
                        wallet.getId(),
                        invoiceDate.getMonthValue(),
                        invoiceDate.getYear(),
                        rebateValue);

                WindowUtils.showSuccessDialog(
                        i18nService.tr(
                                Constants.TranslationKeys.CREDITCARD_DIALOG_INVOICE_PAID_TITLE),
                        i18nService.tr(
                                Constants.TranslationKeys.CREDITCARD_DIALOG_INVOICE_PAID_MESSAGE));

                Stage stage = (Stage) crcNameLabel.getScene().getWindow();
                stage.close();
            } catch (EntityNotFoundException
                    | IllegalArgumentException
                    | MoinexException.InsufficientResourcesException e) {
                WindowUtils.showErrorDialog(
                        i18nService.tr(
                                Constants.TranslationKeys
                                        .CREDITCARD_DIALOG_ERROR_PAYING_INVOICE_TITLE),
                        e.getMessage());
            }
        }

        Stage stage = (Stage) crcNameLabel.getScene().getWindow();
        stage.close();
    }

    @FXML
    private void handleOpenCalculator() {
        WindowUtils.openPopupWindow(
                Constants.CALCULATOR_FXML,
                "Calculator",
                springContext,
                (CalculatorController controller) -> {},
                List.of(() -> calculatorService.updateComponentWithResult(useRebateValueField)));
    }

    private void walletAfterBalance() {
        Wallet wallet = walletComboBox.getValue();

        if (wallet == null) {
            UIUtils.resetLabel(walletAfterBalanceLabel);
            return;
        }

        BigDecimal rebateValue =
                useRebateValueField.getText().isEmpty()
                        ? BigDecimal.ZERO
                        : new BigDecimal(useRebateValueField.getText());

        BigDecimal invoiceAmount =
                creditCardService
                        .getPendingCreditCardPayments(
                                creditCard.getId(),
                                invoiceDate.getMonthValue(),
                                invoiceDate.getYear())
                        .stream()
                        .map(CreditCardPayment::getAmount)
                        .reduce(BigDecimal.ZERO, BigDecimal::add)
                        .subtract(rebateValue);

        if (invoiceAmount.compareTo(BigDecimal.ZERO) < 0) {
            invoiceAmount = BigDecimal.ZERO;
        }

        totalToPayLabel.setText(UIUtils.formatCurrency(invoiceAmount));

        try {
            BigDecimal walletAfterBalanceValue = wallet.getBalance().subtract(invoiceAmount);

            // Set the style according to the balance value after the expense
            if (walletAfterBalanceValue.compareTo(BigDecimal.ZERO) < 0) {
                // Remove old style and add negative style
                UIUtils.setLabelStyle(walletAfterBalanceLabel, Constants.NEGATIVE_BALANCE_STYLE);
            } else {
                // Remove old style and add neutral style
                UIUtils.setLabelStyle(walletAfterBalanceLabel, Constants.NEUTRAL_BALANCE_STYLE);
            }

            walletAfterBalanceLabel.setText(UIUtils.formatCurrency(walletAfterBalanceValue));
        } catch (NumberFormatException e) {
            UIUtils.resetLabel(walletAfterBalanceLabel);
        }
    }

    private void loadWalletsFromDatabase() {
        wallets = walletService.getAllNonArchivedWalletsOrderedByName();
    }

    private void populateWalletComboBox() {
        walletComboBox.getItems().addAll(wallets);
    }

    private void configureComboBoxes() {
        UIUtils.configureComboBox(walletComboBox, Wallet::getName);
    }
}

File: src/main/kotlin/org/moinex/ui/dialog/creditcard/CreditCardInvoicePaymentController.kt
/*
 * Filename: CreditCardInvoicePaymentController.kt (original filename: CreditCardInvoicePaymentController.java)
 * Created on: October 30, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 20/03/2026
 */

package org.moinex.ui.dialog.creditcard

import javafx.fxml.FXML
import javafx.scene.control.ComboBox
import javafx.scene.control.Label
import javafx.scene.control.TextField
import javafx.stage.Stage
import org.moinex.common.constant.Constants
import org.moinex.common.constant.Files
import org.moinex.common.constant.Styles
import org.moinex.common.constant.TranslationKeys
import org.moinex.common.extension.isZero
import org.moinex.common.util.UIUtils
import org.moinex.common.util.WindowUtils
import org.moinex.model.creditcard.CreditCard
import org.moinex.model.dto.CreditCardInvoicePaymentDTO
import org.moinex.model.wallettransaction.Wallet
import org.moinex.service.CalculatorService
import org.moinex.service.PreferencesService
import org.moinex.service.creditcard.CreditCardService
import org.moinex.service.creditcard.RecurringCreditCardDebtService
import org.moinex.service.wallet.WalletService
import org.moinex.ui.common.CalculatorController
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.stereotype.Controller
import java.math.BigDecimal
import java.time.YearMonth

@Controller
class CreditCardInvoicePaymentController(
    private val walletService: WalletService,
    private val creditCardService: CreditCardService,
    private val recurringCreditCardDebtService: RecurringCreditCardDebtService,
    private val calculatorService: CalculatorService,
    private val springContext: ConfigurableApplicationContext,
    private val preferencesService: PreferencesService,
) {
    @FXML
    private lateinit var crcNameLabel: Label

    @FXML
    private lateinit var crcInvoiceDueLabel: Label

    @FXML
    private lateinit var crcInvoiceMonthLabel: Label

    @FXML
    private lateinit var crcAvailableRebateLabel: Label

    @FXML
    private lateinit var crcProjectedAmountLabel: Label

    @FXML
    private lateinit var alreadyPaidValueLabel: Label

    @FXML
    private lateinit var walletAfterBalanceLabel: Label

    @FXML
    private lateinit var walletCurrentBalanceLabel: Label

    @FXML
    private lateinit var totalToPayLabel: Label

    @FXML
    private lateinit var walletComboBox: ComboBox<Wallet>

    @FXML
    private lateinit var useRebateValueField: TextField

    @FXML
    private lateinit var partialAmountField: TextField

    private var wallets: List<Wallet> = emptyList()
    private var creditCard: CreditCard? = null
    private var invoiceDate: YearMonth? = null

    companion object {
        private const val AVAILABLE_REBATE_STYLE = "-fx-text-fill: green;"
        private const val UNAVAILABLE_REBATE_STYLE = "-fx-text-fill: black;"
    }

    fun setCreditCard(
        crc: CreditCard,
        invoiceDate: YearMonth,
    ) {
        this.creditCard = crc
        this.invoiceDate = invoiceDate
        configureComboBoxes()

        crcNameLabel.text = creditCard!!.name

        val invoiceAmount =
            creditCardService.getTotalPendingPaymentsByCreditCardAndMonth(
                creditCard!!.id!!,
                invoiceDate,
            )

        val alreadyPaid =
            creditCardService.getTotalAdvancePaidForInvoice(
                creditCard!!.id!!,
                invoiceDate,
            )

        crcInvoiceDueLabel.text = UIUtils.formatCurrency(invoiceAmount + alreadyPaid)
        totalToPayLabel.text = UIUtils.formatCurrency(invoiceAmount)
        crcInvoiceMonthLabel.text = UIUtils.formatShortMonthYear(invoiceDate)
        crcAvailableRebateLabel.text = UIUtils.formatCurrency(creditCard!!.availableRebate)

        if (alreadyPaid > BigDecimal.ZERO) {
            alreadyPaidValueLabel.text = UIUtils.formatCurrency(alreadyPaid)
        } else {
            alreadyPaidValueLabel.text = "-"
        }

        val projectedAmount =
            recurringCreditCardDebtService
                .getProjectedOccurrencesForMonth(invoiceDate)
                .filter { it.recurringDebt.creditCard.id!! == creditCard!!.id!! }
                .sumOf { it.amount }

        crcProjectedAmountLabel.text = UIUtils.formatCurrency(projectedAmount)

        if (creditCard!!.availableRebate > BigDecimal.ZERO) {
            crcAvailableRebateLabel.style = AVAILABLE_REBATE_STYLE
        } else {
            crcAvailableRebateLabel.style = UNAVAILABLE_REBATE_STYLE
        }

        creditCard!!.defaultBillingWallet?.let { wallet ->
            walletComboBox.value = wallet
            UIUtils.updateWalletBalanceLabelStyle(walletComboBox.value, walletCurrentBalanceLabel)
            walletAfterBalance()
        }
    }

    @FXML
    private fun initialize() {
        configureComboBoxes()
        loadWalletsFromDatabase()
        populateWalletComboBox()

        UIUtils.resetLabel(walletAfterBalanceLabel)
        UIUtils.resetLabel(walletCurrentBalanceLabel)
        UIUtils.resetLabel(crcNameLabel)
        UIUtils.resetLabel(crcInvoiceDueLabel)
        UIUtils.resetLabel(crcInvoiceMonthLabel)

        walletComboBox.setOnAction {
            UIUtils.updateWalletBalanceLabelStyle(walletComboBox.value, walletCurrentBalanceLabel)
            walletAfterBalance()
        }

        useRebateValueField.textProperty().addListener { _, oldValue, newValue ->
            if (!newValue.matches(Regex(Constants.MONETARY_VALUE_REGEX))) {
                useRebateValueField.text = oldValue
            } else {
                walletAfterBalance()
            }
        }

        partialAmountField.textProperty().addListener { _, oldValue, newValue ->
            if (!newValue.matches(Regex(Constants.MONETARY_VALUE_REGEX))) {
                partialAmountField.text = oldValue
            } else {
                walletAfterBalance()
            }
        }
    }

    @FXML
    private fun handleCancel() {
        (crcNameLabel.scene.window as Stage).close()
    }

    @FXML
    private fun handleSave() {
        val wallet = walletComboBox.value

        if (wallet == null) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.CREDITCARD_DIALOG_WALLET_NOT_SELECTED_TITLE),
                preferencesService.translate(TranslationKeys.CREDITCARD_DIALOG_WALLET_NOT_SELECTED_MESSAGE),
            )
            return
        }

        val remainingAmount =
            creditCardService.getTotalPendingPaymentsByCreditCardAndMonth(
                creditCard!!.id!!,
                invoiceDate!!,
            )

        val rebateValue =
            if (useRebateValueField.text.isEmpty()) BigDecimal.ZERO else BigDecimal(useRebateValueField.text)

        if (remainingAmount.isZero()) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.CREDITCARD_DIALOG_INVOICE_ALREADY_PAID_TITLE),
                preferencesService.translate(TranslationKeys.CREDITCARD_DIALOG_INVOICE_ALREADY_PAID_MESSAGE),
            )
            return
        }

        val isPartial = partialAmountField.text.isNotEmpty()

        val amountToPay =
            if (isPartial) BigDecimal(partialAmountField.text) else remainingAmount

        runCatching {
            creditCardService.payInvoice(
                CreditCardInvoicePaymentDTO(
                    creditCardId = creditCard!!.id!!,
                    billingWalletId = wallet.id!!,
                    invoiceDate = invoiceDate!!,
                    amount = amountToPay,
                    rebate = rebateValue,
                ),
            )

            (crcNameLabel.scene.window as Stage).close()
        }.onFailure { e ->
            WindowUtils.showErrorDialog(
                preferencesService.translate(TranslationKeys.CREDITCARD_DIALOG_ERROR_PAYING_INVOICE_TITLE),
                e.message ?: "Unknown error",
            )
        }
    }

    @FXML
    private fun handleOpenCalculator() {
        WindowUtils.openPopupWindow(
            Files.CALCULATOR_FXML,
            preferencesService.translate(TranslationKeys.MAIN_CALCULATOR),
            springContext,
            { _: CalculatorController -> },
            listOf(Runnable { calculatorService.updateComponentWithResult(useRebateValueField) }),
        )
    }

    @FXML
    private fun handleOpenPartialCalculator() {
        WindowUtils.openPopupWindow(
            Files.CALCULATOR_FXML,
            preferencesService.translate(TranslationKeys.MAIN_CALCULATOR),
            springContext,
            { _: CalculatorController -> },
            listOf(Runnable { calculatorService.updateComponentWithResult(partialAmountField) }),
        )
    }

    private fun walletAfterBalance() {
        val wallet = walletComboBox.value

        if (wallet == null) {
            UIUtils.resetLabel(walletAfterBalanceLabel)
            return
        }

        val rebateValue =
            if (useRebateValueField.text.isEmpty()) BigDecimal.ZERO else BigDecimal(useRebateValueField.text)

        val amountToPay =
            if (partialAmountField.text.isNotEmpty()) {
                BigDecimal(partialAmountField.text).subtract(rebateValue).coerceAtLeast(BigDecimal.ZERO)
            } else {
                creditCardService
                    .getTotalPendingPaymentsByCreditCardAndMonth(creditCard!!.id!!, invoiceDate!!)
                    .subtract(rebateValue)
                    .coerceAtLeast(BigDecimal.ZERO)
            }

        totalToPayLabel.text = UIUtils.formatCurrency(amountToPay)

        runCatching {
            val walletAfterBalanceValue = wallet.balance.subtract(amountToPay)

            if (walletAfterBalanceValue < BigDecimal.ZERO) {
                UIUtils.setLabelStyle(walletAfterBalanceLabel, Styles.NEGATIVE_BALANCE_STYLE)
            } else {
                UIUtils.setLabelStyle(walletAfterBalanceLabel, Styles.NEUTRAL_BALANCE_STYLE)
            }

            walletAfterBalanceLabel.text = UIUtils.formatCurrency(walletAfterBalanceValue)
        }.onFailure {
            UIUtils.resetLabel(walletAfterBalanceLabel)
        }
    }

    private fun loadWalletsFromDatabase() {
        wallets = walletService.getAllNonArchivedWalletsOrderedByName()
    }

    private fun populateWalletComboBox() {
        walletComboBox.items.addAll(wallets)
    }

    private fun configureComboBoxes() {
        UIUtils.configureComboBox(walletComboBox, Wallet::name)
    }
}
